rero_ils/modules/serializers/base.py

```diff
@@ -0,0 +1,192 @@
+
+    @staticmethod
+    def preprocess_search_hit(pid, record_hit, links_factory=None, **kwargs):
+        """Prepare a record hit from Elasticsearch for serialization."""
+        record = _JSONSerializer.preprocess_search_hit(
+            pid=pid,
+            record_hit=record_hit,
+            links_factory=links_factory,
+            kwargs=kwargs
+        )
+        if record.get('_explanation'):
+            record[1:] = record.get('_explanation')
+            del record['_explanation']
+        return record
+
+    def serialize_search(self, pid_fetcher, search_result, links=None,
+                         item_links_factory=None, **kwargs):
+        """Serialize a search result.
+
+        :param pid_fetcher: Persistent identifier fetcher.
+        :param search_result: Elasticsearch search result.
+        :param links: Dictionary of links to add to response.
+        :param item_links_factory: Factory function for record links.
+        """
+        results = dict(
+            hits=dict(
+                hits=[
+                    self.transform_search_hit(
+                        pid_fetcher(hit['_id'], hit['_source']),
+                        hit, links_factory=item_links_factory, **kwargs)
+                    for hit in search_result['hits']['hits']
+                ],
+                total=search_result['hits']['total'],
+            ),
+            links=links or {},
+            aggregations=search_result.get('aggregations', dict()),
+        )
+        return json.dumps(
+            self.postprocess_serialize_search(results, pid_fetcher),
+            **self._format_args()
+        )
+
+    @staticmethod
+    def enrich_bucket_with_data(buckets, search_cls, attributes_name):
+        """Complete a bucket by adding new keys based on resource attributes.
+
+        :param buckets: the buckets (aggregation set) to perform.
+        :param search_cls: the search class related to the resource to resolve.
+        :param attributes_name: attributes to load from search data.
+        """
+        attributes_name = attributes_name or []
+        if not isinstance(attributes_name, list):
+            attributes_name = [attributes_name]
+        # search all requested values using search class
+        query = search_cls() \
+            .filter('terms', pid=[term['key'] for term in buckets]) \
+            .source(['pid'] + attributes_name)
+        data = {result.pid: result.to_dict() for result in query.scan()}
+        # complete buckets with data
+        for term in buckets:
+            for attr in attributes_name:
+                if attr in data[term['key']]:
+                    term[attr] = data[term['key']].get(attr)
+
+
+class CachedDataSerializerMixin:
+    """Class to load and cached resources for serialization process."""
+
+    _resources = {}
+
+    def get_resource(self, loader, pid):
+        """Get a resource and store it into the cache if necessary.
```

Comment:
Is the cache expired (or resets) after each serializations ?
If not, I think this is a problem when a resource is updated.

Reply:
The cache is only apply for the object life cycle. As the object is only use for one serialization, it seems good.